### DISCUSSION TITLE
Increasing the base font size when using rdMolDraw2D

### DISCUSSION THREAD

**Discussion by rachelnwalker** [rdkit/rdkit#4652] [Q&A]
I have this basic function to draw a molecule, but I would like to make the font sizes larger when possible. Here is the function:
```
def draw_mol(mol):
    drawer = rdMolDraw2D.MolDraw2DSVG(400, 400)
    draw_options = drawer.drawOptions()
    draw_options.minFontSize = -1
    draw_options.maxFontSize = -1
    draw_options.scaleBondWidth = True
    drawer.DrawMolecule(mol)
    drawer.FinishDrawing()
    print(f'font size: {drawer.FontSize()}')
    return drawer.GetDrawingText()
```
And here is an example output:
![image](https://user-images.githubusercontent.com/39069546/139298364-593d15b0-8ba7-4349-b05f-13c23ba2ec18.png)

I would like to increase the font size for mols scaled like the one on the right. I tried some of the drawing options, but the only way I could get the font size to increase was to increase the minimum font size — which doesn’t generalize well to large structures:
![image](https://user-images.githubusercontent.com/39069546/139298556-242f918a-609e-44c8-9ea6-619f118532d4.png)

Is there a way to increase the 'base' font size without requiring a minimum? I sort of have a work around will scale the minimum font size based on the how a molecule’s coordinates fits into a window, but I'm curious whether there is a drawing option to do that. 

I also looked a little bit at some of the C++ DrawingText code, and it looks like there is a 'font size' that is hard coded and scaled during image generation: https://github.com/rdkit/rdkit/blob/b208da471f8edc88e07c77ed7d7868649ac75100/Code/GraphMol/MolDraw2D/DrawText.h#L40

**Comment by DavidACosgrove** [comment]
Hi @rachelnwalker,
You're not the first person to ask for this.  Last time I looked at it, I convinced myself it was not trivial.  I will have another poke about and see what can be done.  It won't be soon, but I will do my best not to forget it.  Feel free to ping me if you think it's been too long!  Since this is a feature request, I guess the deadline to do something would be the next release.  
Regards,
Dave

**Comment by greglandrum** [comment]
@DavidACosgrove one thing that occurred to me here and which may kill multiple birds with one stone: if we (where by "we" I think I mean "me") added a new drawing mode which allowed the user to specify the canvas size as -1,-1 and set the font size and number of pixels per "angstrom", we could pretty easily modify the existing scaling code, have it calculate the required canvas size, and provide consistently scaled structures.

Ok, that doesn't actually solve the whole problem that Rachel is having, but I think it'd still be useful and would allow configuration of things like publication modes which have the appropriate settings for journals.

**Comment by DavidACosgrove** [comment]
@greglandrum I think that makes sense.
I like the "we" == "me" part!  I hope you don't end up hating the way the drawing currently swaps between the 2 drawing coordinate frames and cursing the author!  I did find the freetype coordinate system awkward, not least its ignoring of the normal graphics convention that +ve y goes down the page.
Let me know if you want to discuss this offline.

**Comment by DavidACosgrove** [comment]
Hi @greglandrum, @rachelnwalker,

Whilst attempting to distract myself from a particularly unpleasant session in the gym, I have remembered why a minimum font size is difficult.  The user generally wants to specify the minimum size as a function of the final drawing - no less than 10 pixels high, or something.  We pick a font size that looks like a good proportion of the bond lengths and to an arbitrary scale, truncating the lines for bonds so they don't go over the characters in the drawing.  The picture is then scaled to fit the required final size.  If that final scale makes the characters too small for the minimum font size, the fonts and the rest of the drawing have to be scaled differently or all the ends of the bonds are wrong so they start impinging on the characters.  We are not set up to do that.  This may just be a feature of the organic way the drawing code has evolved over the years, but as it stands that is the problem.  

I haven't thought this through thoroughly, but it feels as though you almost need to draw the characters first at the required scale, and then put the lines in, possibly at a different scale, clipping them to deal with required font size.  An issue there, though, is that the final scale needs to accommodate the size of the characters so they don't get clipped off the edges of the drawing when the atomic symbol is at an extreme of the molecule.  I hope that made sense - it's the sort of thing that is a lot easier to explain with the aid of a whiteboard and some hand-waving.  I think given where we are, there would have to be a significant refactoring of the code to allow for the minimum font size, and I'm pretty sure I have neither the stamina nor the time for that at the moment.  I think all the pieces are there, but they will need re-arranging and no doubt re-working a bit as well.

Regards,
Dave

**Comment by DavidACosgrove** [comment]
Hi @rachelnwalker.  @greglandrum is funding me to do some re-factoring of the drawing code.  I think I have produced a good solution to the minimum font size, and in his last round of fettling the code he also fixed the base font size, I think.  It isn't as handy as setting the font size to so many pixels, but it does work.  Could you send your input structures above and I will have a play and make sure it is behaving sensibly, and then you can pass judgement on the results?  It seems like a good opportunity to get this sorted to your satisfaction.
Thanks,
Dave

**Reply by radusuciu** [reply]
This would definitely be useful. Do you still need structures to test?

**Reply by rachelnwalker** [reply]
@DavidACosgrove sorry for taking a bit to get back to you! Here is a text file with around 35 molblocks I was testing on originally -- each molblock is the same as previous with ~10 less atoms so it should help to visualize how the font size changes for different sized structures. Hopefully this is still useful -- thanks!

[font_size_scaling_examples.txt](https://github.com/rdkit/rdkit/files/7949226/font_size_scaling_examples.txt)

**Comment by DavidACosgrove** [comment]
Hi @rachelnwalker.  Many thanks.  Very timely, in fact.

**Comment by DavidACosgrove** [comment]
Hi @rachelnwalker.  I can confirm that in the forthcoming enhancements to the molecule drawers, setting baseFontSize in drawOptions does affect the drawing in a useful manner.  Hopefully that'll come in the March release.  A key problem is that baseFontSize isn't exposed to the Python wrapper at present.  That feels like it might be a bug fix that could be rolled out earlier in which case you would be able to use it sooner.  If @greglandrum agrees this is worthy of a bug release, I'll submit a PR - it's a 2-liner.
